{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "SHORTEN = {\n",
    "    \"Artery-Aorta\":\"Artery.A\"     ,\n",
    "    \"Artery-Tibial\": \"Artery.T\",\n",
    "    \"Adipose-Subcutaneous\": \"Adipose S.\",    \n",
    "    \"Adipose-Visceral(Omentum)\":\"Adipose V.\",\n",
    "    \"Brain-Caudate(basalganglia)\":\"B.Caudate\"   , \n",
    "    \"Brain-Cerebellum\":\"B.Cerebellum\",\n",
    "    \"Cells-Transformedfibroblasts\": \"Fibroblast\",\n",
    "    \"Esophagus-Mucosa\": \"E.Mucosa\",\n",
    "    \"Esophagus-Muscularis\":\"E.Muscularis\",\n",
    "    \"Heart-LeftVentricle\":\"Ventricule\",\n",
    "    \"Lung\": \"Lung\",\n",
    "    \"Muscle-Skeletal\": \"Muscle\",\n",
    "    \"Nerve-Tibial\":\"Nerve\",\n",
    "    \"Skin-NotSunExposed(Suprapubic)\": \"Skin.Unexposed\",\n",
    "    \"Skin-SunExposed(Lowerleg)\":\"Skin.Leg\",\n",
    "    \"Thyroid\":\"Thyroid\",\n",
    "    \"WholeBlood\": \"Blood\",\n",
    "}\n",
    "\n",
    "import math\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from collections import defaultdict\n",
    "import subprocess\n",
    "\n",
    "tissu = sorted([x for x in SHORTEN.keys()])\n",
    "CHR = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,'X']\n",
    "Gene_table = pd.read_csv('/storage/szfeupe/Runs/GTEx_estr/FEATURES/Genes_only_table', sep='\\t')\n",
    "Motifs = pd.read_csv('/storage/resources/dbase/human/hg19/hg19.hipstr_reference_withmotif.bed', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Adipose-Subcutaneous\n",
      "Adipose-Visceral(Omentum)\n",
      "Artery-Aorta\n",
      "Artery-Tibial\n",
      "Brain-Caudate(basalganglia)\n",
      "Brain-Cerebellum\n",
      "Cells-Transformedfibroblasts\n",
      "Esophagus-Mucosa\n",
      "Esophagus-Muscularis\n",
      "Heart-LeftVentricle\n",
      "Lung\n",
      "Muscle-Skeletal\n",
      "Nerve-Tibial\n",
      "Skin-NotSunExposed(Suprapubic)\n",
      "Skin-SunExposed(Lowerleg)\n",
      "Thyroid\n",
      "WholeBlood\n",
      "END!\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Samples</th>\n",
       "      <th>Num.genes</th>\n",
       "      <th>Num.estrs</th>\n",
       "      <th>Num.BestCaviar</th>\n",
       "      <th>eSTR.CAVIAR</th>\n",
       "      <th>ANOVA.pass</th>\n",
       "      <th>ANOVA.STR+SNP</th>\n",
       "      <th>eSTR+SNP.anova.pass</th>\n",
       "      <th>e.ANOVA.CAVIAR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Nerve</th>\n",
       "      <td>265</td>\n",
       "      <td>12567</td>\n",
       "      <td>1681</td>\n",
       "      <td>258</td>\n",
       "      <td>111</td>\n",
       "      <td>407</td>\n",
       "      <td>690</td>\n",
       "      <td>276</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E.Muscularis</th>\n",
       "      <td>214</td>\n",
       "      <td>12432</td>\n",
       "      <td>1028</td>\n",
       "      <td>220</td>\n",
       "      <td>83</td>\n",
       "      <td>255</td>\n",
       "      <td>611</td>\n",
       "      <td>187</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Adipose S.</th>\n",
       "      <td>270</td>\n",
       "      <td>12432</td>\n",
       "      <td>1353</td>\n",
       "      <td>248</td>\n",
       "      <td>105</td>\n",
       "      <td>304</td>\n",
       "      <td>573</td>\n",
       "      <td>206</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Thyroid</th>\n",
       "      <td>262</td>\n",
       "      <td>12616</td>\n",
       "      <td>1523</td>\n",
       "      <td>249</td>\n",
       "      <td>99</td>\n",
       "      <td>367</td>\n",
       "      <td>647</td>\n",
       "      <td>251</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Artery.T</th>\n",
       "      <td>276</td>\n",
       "      <td>12140</td>\n",
       "      <td>1257</td>\n",
       "      <td>242</td>\n",
       "      <td>102</td>\n",
       "      <td>315</td>\n",
       "      <td>562</td>\n",
       "      <td>220</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Muscle</th>\n",
       "      <td>343</td>\n",
       "      <td>11970</td>\n",
       "      <td>1273</td>\n",
       "      <td>245</td>\n",
       "      <td>98</td>\n",
       "      <td>301</td>\n",
       "      <td>442</td>\n",
       "      <td>178</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E.Mucosa</th>\n",
       "      <td>255</td>\n",
       "      <td>12474</td>\n",
       "      <td>1228</td>\n",
       "      <td>244</td>\n",
       "      <td>91</td>\n",
       "      <td>276</td>\n",
       "      <td>563</td>\n",
       "      <td>190</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Adipose V.</th>\n",
       "      <td>193</td>\n",
       "      <td>12528</td>\n",
       "      <td>693</td>\n",
       "      <td>288</td>\n",
       "      <td>81</td>\n",
       "      <td>157</td>\n",
       "      <td>560</td>\n",
       "      <td>111</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Skin.Leg</th>\n",
       "      <td>297</td>\n",
       "      <td>12606</td>\n",
       "      <td>1462</td>\n",
       "      <td>233</td>\n",
       "      <td>106</td>\n",
       "      <td>321</td>\n",
       "      <td>514</td>\n",
       "      <td>198</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Fibroblast</th>\n",
       "      <td>225</td>\n",
       "      <td>11897</td>\n",
       "      <td>1108</td>\n",
       "      <td>245</td>\n",
       "      <td>80</td>\n",
       "      <td>239</td>\n",
       "      <td>539</td>\n",
       "      <td>161</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ventricule</th>\n",
       "      <td>199</td>\n",
       "      <td>12045</td>\n",
       "      <td>605</td>\n",
       "      <td>248</td>\n",
       "      <td>64</td>\n",
       "      <td>128</td>\n",
       "      <td>489</td>\n",
       "      <td>102</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lung</th>\n",
       "      <td>259</td>\n",
       "      <td>12684</td>\n",
       "      <td>1065</td>\n",
       "      <td>280</td>\n",
       "      <td>86</td>\n",
       "      <td>236</td>\n",
       "      <td>596</td>\n",
       "      <td>149</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Blood</th>\n",
       "      <td>336</td>\n",
       "      <td>11571</td>\n",
       "      <td>883</td>\n",
       "      <td>245</td>\n",
       "      <td>78</td>\n",
       "      <td>214</td>\n",
       "      <td>431</td>\n",
       "      <td>142</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Skin.Unexposed</th>\n",
       "      <td>209</td>\n",
       "      <td>12661</td>\n",
       "      <td>840</td>\n",
       "      <td>257</td>\n",
       "      <td>67</td>\n",
       "      <td>161</td>\n",
       "      <td>530</td>\n",
       "      <td>110</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B.Cerebellum</th>\n",
       "      <td>107</td>\n",
       "      <td>12591</td>\n",
       "      <td>498</td>\n",
       "      <td>267</td>\n",
       "      <td>53</td>\n",
       "      <td>89</td>\n",
       "      <td>648</td>\n",
       "      <td>79</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Artery.A</th>\n",
       "      <td>191</td>\n",
       "      <td>12338</td>\n",
       "      <td>821</td>\n",
       "      <td>247</td>\n",
       "      <td>63</td>\n",
       "      <td>160</td>\n",
       "      <td>589</td>\n",
       "      <td>117</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B.Caudate</th>\n",
       "      <td>108</td>\n",
       "      <td>12650</td>\n",
       "      <td>204</td>\n",
       "      <td>320</td>\n",
       "      <td>28</td>\n",
       "      <td>38</td>\n",
       "      <td>696</td>\n",
       "      <td>31</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Samples  Num.genes  Num.estrs  Num.BestCaviar  eSTR.CAVIAR  \\\n",
       "Nerve               265      12567       1681             258          111   \n",
       "E.Muscularis        214      12432       1028             220           83   \n",
       "Adipose S.          270      12432       1353             248          105   \n",
       "Thyroid             262      12616       1523             249           99   \n",
       "Artery.T            276      12140       1257             242          102   \n",
       "Muscle              343      11970       1273             245           98   \n",
       "E.Mucosa            255      12474       1228             244           91   \n",
       "Adipose V.          193      12528        693             288           81   \n",
       "Skin.Leg            297      12606       1462             233          106   \n",
       "Fibroblast          225      11897       1108             245           80   \n",
       "Ventricule          199      12045        605             248           64   \n",
       "Lung                259      12684       1065             280           86   \n",
       "Blood               336      11571        883             245           78   \n",
       "Skin.Unexposed      209      12661        840             257           67   \n",
       "B.Cerebellum        107      12591        498             267           53   \n",
       "Artery.A            191      12338        821             247           63   \n",
       "B.Caudate           108      12650        204             320           28   \n",
       "\n",
       "                ANOVA.pass  ANOVA.STR+SNP  eSTR+SNP.anova.pass  e.ANOVA.CAVIAR  \n",
       "Nerve                  407            690                  276              56  \n",
       "E.Muscularis           255            611                  187              52  \n",
       "Adipose S.             304            573                  206              49  \n",
       "Thyroid                367            647                  251              45  \n",
       "Artery.T               315            562                  220              44  \n",
       "Muscle                 301            442                  178              43  \n",
       "E.Mucosa               276            563                  190              41  \n",
       "Adipose V.             157            560                  111              38  \n",
       "Skin.Leg               321            514                  198              36  \n",
       "Fibroblast             239            539                  161              32  \n",
       "Ventricule             128            489                  102              32  \n",
       "Lung                   236            596                  149              32  \n",
       "Blood                  214            431                  142              32  \n",
       "Skin.Unexposed         161            530                  110              29  \n",
       "B.Cerebellum            89            648                   79              24  \n",
       "Artery.A               160            589                  117              24  \n",
       "B.Caudate               38            696                   31              15  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Table0 = defaultdict(list)\n",
    "for Tissue in tissu:\n",
    "    print(Tissue)\n",
    "\n",
    "    LR = '/storage/szfeupe/Runs/650GTEx_estr/Analysis_by_Tissue/'+Tissue+'/PQValues'\n",
    "    ANOV = '/storage/szfeupe/Runs/650GTEx_estr/Analysis_by_Tissue/'+Tissue+'/HH/anova_wg.csv'\n",
    "    CAUSE = '/storage/szfeupe/Runs/650GTEx_estr/Analysis_by_Tissue/'+Tissue+'/HH/caviar.out'\n",
    "    samples = '/storage/szfeupe/Runs/650GTEx_estr/Analysis_by_Tissue/'+Tissue+'/Corr_Expr.csv'\n",
    "\n",
    "    anova = pd.read_csv(ANOV, sep=',')\n",
    "    anova = anova[['chrom','gene','str.start','anova_pval','estr_fdr','esnp_fdr','delta_bic','delta_aic']]\n",
    "    lr = pd.read_csv(LR,sep='\\t')\n",
    "    Cs=[]\n",
    "    for i in CHR:\n",
    "        C=pd.read_csv(CAUSE+str(i), sep='\\t')\n",
    "        Cs.append(C)\n",
    "    causal=Cs[0]\n",
    "    for i in range(1,23,1):\n",
    "        causal=causal.append(Cs[i])   \n",
    "    causal['best.str.start'] = causal['best.str.start'].astype(int)\n",
    "    caviarbest=[x for x in list(causal['top.variant']) if 'STR_' in x]\n",
    "    estrs = lr.loc[lr['qvalue']<=0.1]\n",
    "    Table0['Num.genes'].append(lr.shape[0])\n",
    "    Table0['Num.estrs'].append(estrs.shape[0])\n",
    "    anovaP = anova.loc[anova['anova_pval']<=0.05]  #anova pass\n",
    "    anovaPe = anovaP.loc[anovaP['gene'].isin(list(estrs['gene']))]  #estrs + anova pass\n",
    "    anovaSTR = anova.loc[(anova['delta_aic']>0) & (anova['delta_bic']>0)] #snp+str prefered\n",
    "    X = anovaSTR.loc[anovaSTR['gene'].isin(list(anovaPe['gene']))]  #estrs + snp+STR + pass\n",
    "    Table0['ANOVA.pass'].append(anovaPe.shape[0])\n",
    "    Table0['Num.BestCaviar'].append(len(caviarbest))\n",
    "    cmd = len(open(samples,'r').readlines())\n",
    "    Table0['Samples'].append (int(cmd)-1)\n",
    "    Table0['ANOVA.STR+SNP'].append(anovaSTR.shape[0])\n",
    "    Table0['eSTR+SNP.anova.pass'].append(X.shape[0])\n",
    "    Table0['e.ANOVA.CAVIAR'].append(X.loc[X['gene'].isin(list(causal.loc[causal['top.variant'].isin(caviarbest)]['gene']))].shape[0])\n",
    "    Table0['eSTR.CAVIAR'].append(estrs.loc[estrs['gene'].isin(list(causal.loc[causal['top.variant'].isin(caviarbest)]['gene']))].shape[0])\n",
    "\n",
    " \n",
    "Table0['Tissues']=tissu\n",
    "Table = pd.DataFrame(Table0)\n",
    "Table.index =[SHORTEN[x] for x in list( Table['Tissues'])]\n",
    "del Table['Tissues']\n",
    "Table = Table[['Samples','Num.genes','Num.estrs','Num.BestCaviar','eSTR.CAVIAR','ANOVA.pass','ANOVA.STR+SNP','eSTR+SNP.anova.pass','e.ANOVA.CAVIAR']]\n",
    "print('END!')\n",
    "\n",
    "Table.sort_values(\"e.ANOVA.CAVIAR\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>gene</th>\n",
       "      <th>str.start</th>\n",
       "      <th>anova_pval</th>\n",
       "      <th>estr_fdr</th>\n",
       "      <th>esnp_fdr</th>\n",
       "      <th>delta_bic</th>\n",
       "      <th>delta_aic</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>459</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000134247.9</td>\n",
       "      <td>117556808</td>\n",
       "      <td>0.004327</td>\n",
       "      <td>0.087358</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.589801</td>\n",
       "      <td>6.242290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>582</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000084636.13</td>\n",
       "      <td>32044259</td>\n",
       "      <td>0.000474</td>\n",
       "      <td>0.083936</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>6.798188</td>\n",
       "      <td>10.377918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>918</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000197965.7</td>\n",
       "      <td>167737092</td>\n",
       "      <td>0.001980</td>\n",
       "      <td>0.016046</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>3.947508</td>\n",
       "      <td>7.677608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>976</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000135750.10</td>\n",
       "      <td>233797970</td>\n",
       "      <td>0.008308</td>\n",
       "      <td>0.010981</td>\n",
       "      <td>0.615988</td>\n",
       "      <td>1.291539</td>\n",
       "      <td>5.044112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1814</th>\n",
       "      <td>chr2</td>\n",
       "      <td>ENSG00000162999.8</td>\n",
       "      <td>184035958</td>\n",
       "      <td>0.005452</td>\n",
       "      <td>0.029353</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.013252</td>\n",
       "      <td>5.806266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2211</th>\n",
       "      <td>chr3</td>\n",
       "      <td>ENSG00000174564.8</td>\n",
       "      <td>136642402</td>\n",
       "      <td>0.000570</td>\n",
       "      <td>0.023987</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>6.243729</td>\n",
       "      <td>10.002631</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2769</th>\n",
       "      <td>chr4</td>\n",
       "      <td>ENSG00000163138.14</td>\n",
       "      <td>20607320</td>\n",
       "      <td>0.000522</td>\n",
       "      <td>0.023552</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>6.514626</td>\n",
       "      <td>10.181053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4131</th>\n",
       "      <td>chr6</td>\n",
       "      <td>ENSG00000196230.8</td>\n",
       "      <td>30704959</td>\n",
       "      <td>0.002483</td>\n",
       "      <td>0.016415</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>3.721534</td>\n",
       "      <td>7.278362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4959</th>\n",
       "      <td>chr8</td>\n",
       "      <td>ENSG00000181195.6</td>\n",
       "      <td>57438742</td>\n",
       "      <td>0.005048</td>\n",
       "      <td>0.077194</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.311231</td>\n",
       "      <td>5.960205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5611</th>\n",
       "      <td>chr10</td>\n",
       "      <td>ENSG00000148848.10</td>\n",
       "      <td>127831854</td>\n",
       "      <td>0.000120</td>\n",
       "      <td>0.009128</td>\n",
       "      <td>0.905767</td>\n",
       "      <td>9.196705</td>\n",
       "      <td>12.961896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6168</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSG00000175274.14</td>\n",
       "      <td>44972318</td>\n",
       "      <td>0.005137</td>\n",
       "      <td>0.028650</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.230131</td>\n",
       "      <td>5.923863</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6196</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSG00000137714.2</td>\n",
       "      <td>110224583</td>\n",
       "      <td>0.004272</td>\n",
       "      <td>0.032160</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.564174</td>\n",
       "      <td>6.261268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6460</th>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSG00000168092.9</td>\n",
       "      <td>116927253</td>\n",
       "      <td>0.010591</td>\n",
       "      <td>0.024098</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.906015</td>\n",
       "      <td>4.609797</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6613</th>\n",
       "      <td>chr12</td>\n",
       "      <td>ENSG00000165966.10</td>\n",
       "      <td>41710259</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>0.004461</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>11.692912</td>\n",
       "      <td>15.366236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7073</th>\n",
       "      <td>chr12</td>\n",
       "      <td>ENSG00000135111.10</td>\n",
       "      <td>115115826</td>\n",
       "      <td>0.003447</td>\n",
       "      <td>0.044305</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>3.007964</td>\n",
       "      <td>6.660453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7196</th>\n",
       "      <td>chr13</td>\n",
       "      <td>ENSG00000136156.8</td>\n",
       "      <td>48890856</td>\n",
       "      <td>0.000674</td>\n",
       "      <td>0.009891</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>6.083469</td>\n",
       "      <td>9.707487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7635</th>\n",
       "      <td>chr14</td>\n",
       "      <td>ENSG00000100593.13</td>\n",
       "      <td>78000497</td>\n",
       "      <td>0.002382</td>\n",
       "      <td>0.076574</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>3.781646</td>\n",
       "      <td>7.353800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7827</th>\n",
       "      <td>chr15</td>\n",
       "      <td>ENSG00000166839.12</td>\n",
       "      <td>65117067</td>\n",
       "      <td>0.001990</td>\n",
       "      <td>0.004127</td>\n",
       "      <td>0.001135</td>\n",
       "      <td>3.927641</td>\n",
       "      <td>7.667434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8293</th>\n",
       "      <td>chr16</td>\n",
       "      <td>ENSG00000213398.3</td>\n",
       "      <td>68014740</td>\n",
       "      <td>0.001765</td>\n",
       "      <td>0.004127</td>\n",
       "      <td>0.011501</td>\n",
       "      <td>4.309006</td>\n",
       "      <td>7.907428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8359</th>\n",
       "      <td>chr16</td>\n",
       "      <td>ENSG00000125149.7</td>\n",
       "      <td>67142564</td>\n",
       "      <td>0.009623</td>\n",
       "      <td>0.029552</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.995871</td>\n",
       "      <td>4.776615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8462</th>\n",
       "      <td>chr16</td>\n",
       "      <td>ENSG00000090905.13</td>\n",
       "      <td>24740990</td>\n",
       "      <td>0.000328</td>\n",
       "      <td>0.004127</td>\n",
       "      <td>0.460932</td>\n",
       "      <td>7.379731</td>\n",
       "      <td>11.059904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8921</th>\n",
       "      <td>chr17</td>\n",
       "      <td>ENSG00000004939.9</td>\n",
       "      <td>42324032</td>\n",
       "      <td>0.017532</td>\n",
       "      <td>0.079769</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.090529</td>\n",
       "      <td>3.714547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9459</th>\n",
       "      <td>chr18</td>\n",
       "      <td>ENSG00000154889.12</td>\n",
       "      <td>11968400</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.004127</td>\n",
       "      <td>0.021429</td>\n",
       "      <td>16.224681</td>\n",
       "      <td>19.944992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9466</th>\n",
       "      <td>chr18</td>\n",
       "      <td>ENSG00000134490.9</td>\n",
       "      <td>20848751</td>\n",
       "      <td>0.000418</td>\n",
       "      <td>0.018160</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>6.799743</td>\n",
       "      <td>10.586640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9900</th>\n",
       "      <td>chr19</td>\n",
       "      <td>ENSG00000105552.10</td>\n",
       "      <td>49275684</td>\n",
       "      <td>0.000333</td>\n",
       "      <td>0.078002</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>7.228849</td>\n",
       "      <td>11.015747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960</th>\n",
       "      <td>chr19</td>\n",
       "      <td>ENSG00000051128.14</td>\n",
       "      <td>18954527</td>\n",
       "      <td>0.004505</td>\n",
       "      <td>0.043446</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.463306</td>\n",
       "      <td>6.163750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10127</th>\n",
       "      <td>chr19</td>\n",
       "      <td>ENSG00000141858.7</td>\n",
       "      <td>14289998</td>\n",
       "      <td>0.000093</td>\n",
       "      <td>0.010158</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>9.684648</td>\n",
       "      <td>13.449839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10435</th>\n",
       "      <td>chr19</td>\n",
       "      <td>ENSG00000244165.1</td>\n",
       "      <td>10214984</td>\n",
       "      <td>0.000125</td>\n",
       "      <td>0.029981</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>9.291538</td>\n",
       "      <td>12.911939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11196</th>\n",
       "      <td>chr21</td>\n",
       "      <td>ENSG00000160213.5</td>\n",
       "      <td>45196326</td>\n",
       "      <td>0.000166</td>\n",
       "      <td>0.004127</td>\n",
       "      <td>0.000829</td>\n",
       "      <td>8.545121</td>\n",
       "      <td>12.335082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11249</th>\n",
       "      <td>chr22</td>\n",
       "      <td>ENSG00000100325.10</td>\n",
       "      <td>30234193</td>\n",
       "      <td>0.015050</td>\n",
       "      <td>0.005967</td>\n",
       "      <td>0.741800</td>\n",
       "      <td>0.168475</td>\n",
       "      <td>3.973610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11310</th>\n",
       "      <td>chr22</td>\n",
       "      <td>ENSG00000100345.16</td>\n",
       "      <td>36577729</td>\n",
       "      <td>0.000766</td>\n",
       "      <td>0.028650</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>5.834749</td>\n",
       "      <td>9.465961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11338</th>\n",
       "      <td>chr22</td>\n",
       "      <td>ENSG00000100417.7</td>\n",
       "      <td>41940355</td>\n",
       "      <td>0.014727</td>\n",
       "      <td>0.059324</td>\n",
       "      <td>0.733374</td>\n",
       "      <td>0.223173</td>\n",
       "      <td>4.013133</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom                gene  str.start  anova_pval  estr_fdr  esnp_fdr  \\\n",
       "459     chr1   ENSG00000134247.9  117556808    0.004327  0.087358  1.000000   \n",
       "582     chr1  ENSG00000084636.13   32044259    0.000474  0.083936  1.000000   \n",
       "918     chr1   ENSG00000197965.7  167737092    0.001980  0.016046  1.000000   \n",
       "976     chr1  ENSG00000135750.10  233797970    0.008308  0.010981  0.615988   \n",
       "1814    chr2   ENSG00000162999.8  184035958    0.005452  0.029353  1.000000   \n",
       "2211    chr3   ENSG00000174564.8  136642402    0.000570  0.023987  1.000000   \n",
       "2769    chr4  ENSG00000163138.14   20607320    0.000522  0.023552  1.000000   \n",
       "4131    chr6   ENSG00000196230.8   30704959    0.002483  0.016415  1.000000   \n",
       "4959    chr8   ENSG00000181195.6   57438742    0.005048  0.077194  1.000000   \n",
       "5611   chr10  ENSG00000148848.10  127831854    0.000120  0.009128  0.905767   \n",
       "6168   chr11  ENSG00000175274.14   44972318    0.005137  0.028650  1.000000   \n",
       "6196   chr11   ENSG00000137714.2  110224583    0.004272  0.032160  1.000000   \n",
       "6460   chr11   ENSG00000168092.9  116927253    0.010591  0.024098  1.000000   \n",
       "6613   chr12  ENSG00000165966.10   41710259    0.000034  0.004461  1.000000   \n",
       "7073   chr12  ENSG00000135111.10  115115826    0.003447  0.044305  1.000000   \n",
       "7196   chr13   ENSG00000136156.8   48890856    0.000674  0.009891  1.000000   \n",
       "7635   chr14  ENSG00000100593.13   78000497    0.002382  0.076574  1.000000   \n",
       "7827   chr15  ENSG00000166839.12   65117067    0.001990  0.004127  0.001135   \n",
       "8293   chr16   ENSG00000213398.3   68014740    0.001765  0.004127  0.011501   \n",
       "8359   chr16   ENSG00000125149.7   67142564    0.009623  0.029552  1.000000   \n",
       "8462   chr16  ENSG00000090905.13   24740990    0.000328  0.004127  0.460932   \n",
       "8921   chr17   ENSG00000004939.9   42324032    0.017532  0.079769  1.000000   \n",
       "9459   chr18  ENSG00000154889.12   11968400    0.000003  0.004127  0.021429   \n",
       "9466   chr18   ENSG00000134490.9   20848751    0.000418  0.018160  1.000000   \n",
       "9900   chr19  ENSG00000105552.10   49275684    0.000333  0.078002  1.000000   \n",
       "9960   chr19  ENSG00000051128.14   18954527    0.004505  0.043446  1.000000   \n",
       "10127  chr19   ENSG00000141858.7   14289998    0.000093  0.010158  1.000000   \n",
       "10435  chr19   ENSG00000244165.1   10214984    0.000125  0.029981  1.000000   \n",
       "11196  chr21   ENSG00000160213.5   45196326    0.000166  0.004127  0.000829   \n",
       "11249  chr22  ENSG00000100325.10   30234193    0.015050  0.005967  0.741800   \n",
       "11310  chr22  ENSG00000100345.16   36577729    0.000766  0.028650  1.000000   \n",
       "11338  chr22   ENSG00000100417.7   41940355    0.014727  0.059324  0.733374   \n",
       "\n",
       "       delta_bic  delta_aic  \n",
       "459     2.589801   6.242290  \n",
       "582     6.798188  10.377918  \n",
       "918     3.947508   7.677608  \n",
       "976     1.291539   5.044112  \n",
       "1814    2.013252   5.806266  \n",
       "2211    6.243729  10.002631  \n",
       "2769    6.514626  10.181053  \n",
       "4131    3.721534   7.278362  \n",
       "4959    2.311231   5.960205  \n",
       "5611    9.196705  12.961896  \n",
       "6168    2.230131   5.923863  \n",
       "6196    2.564174   6.261268  \n",
       "6460    0.906015   4.609797  \n",
       "6613   11.692912  15.366236  \n",
       "7073    3.007964   6.660453  \n",
       "7196    6.083469   9.707487  \n",
       "7635    3.781646   7.353800  \n",
       "7827    3.927641   7.667434  \n",
       "8293    4.309006   7.907428  \n",
       "8359    0.995871   4.776615  \n",
       "8462    7.379731  11.059904  \n",
       "8921    0.090529   3.714547  \n",
       "9459   16.224681  19.944992  \n",
       "9466    6.799743  10.586640  \n",
       "9900    7.228849  11.015747  \n",
       "9960    2.463306   6.163750  \n",
       "10127   9.684648  13.449839  \n",
       "10435   9.291538  12.911939  \n",
       "11196   8.545121  12.335082  \n",
       "11249   0.168475   3.973610  \n",
       "11310   5.834749   9.465961  \n",
       "11338   0.223173   4.013133  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#q = pd.read_csv('/storage/szfeupe/Runs/650GTEx_estr/Analysis_by_Tissue/LR_SummaryTest_Table.tsv',sep='\\t')\n",
    "#q.loc[q['gene']=='ENSG00000160213.5']\n",
    "#100*149/2436\n",
    "X.loc[X['gene'].isin(list(causal.loc[causal['top.variant'].isin(caviarbest)]['gene']))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Number of egenes...  (2436)\n",
    "Nbr of eGenes in more than 1 tissues  (1029)\n",
    "egenes in at least 10 tissues  (149)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
