{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Adipose-Subcutaneous  variants ...\n",
      "(87891, 14)   to   (14712, 14)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/lib/python3.4/site-packages/ipykernel/__main__.py:59: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14712  total tests...  2872  pvalues were reduced to 1\n",
      "1789 \t gene level qval<=0.1\n",
      "4516 \t locus level qval<=0.1\n",
      "2027 \t qval<0.01\n",
      "\n",
      "Adipose-Visceral  variants ...\n",
      "(87912, 14)   to   (14803, 14)\n",
      "14803  total tests...  3278  pvalues were reduced to 1\n",
      "888 \t gene level qval<=0.1\n",
      "2017 \t locus level qval<=0.1\n",
      "796 \t qval<0.01\n",
      "\n",
      "Artery-Aorta  variants ...\n",
      "(86353, 14)   to   (14609, 14)\n",
      "14609  total tests...  3076  pvalues were reduced to 1\n",
      "1139 \t gene level qval<=0.1\n",
      "2617 \t locus level qval<=0.1\n",
      "1118 \t qval<0.01\n",
      "\n",
      "Artery-Tibial  variants ...\n",
      "(85366, 14)   to   (14372, 14)\n",
      "14372  total tests...  2722  pvalues were reduced to 1\n",
      "1689 \t gene level qval<=0.1\n",
      "4063 \t locus level qval<=0.1\n",
      "1780 \t qval<0.01\n",
      "\n",
      "Brain-Caudate  variants ...\n",
      "(87076, 14)   to   (14939, 14)\n",
      "14939  total tests...  3680  pvalues were reduced to 1\n",
      "251 \t gene level qval<=0.1\n",
      "488 \t locus level qval<=0.1\n",
      "0 \t qval<0.01\n",
      "\n",
      "Brain-Cerebellum  variants ...\n",
      "(87179, 14)   to   (14858, 14)\n",
      "14858  total tests...  3391  pvalues were reduced to 1\n",
      "643 \t gene level qval<=0.1\n",
      "1340 \t locus level qval<=0.1\n",
      "438 \t qval<0.01\n",
      "\n",
      "Cells-Transformedfibroblasts  variants ...\n",
      "(83075, 14)   to   (14072, 14)\n",
      "14072  total tests...  2818  pvalues were reduced to 1\n",
      "1518 \t gene level qval<=0.1\n",
      "3566 \t locus level qval<=0.1\n",
      "1534 \t qval<0.01\n",
      "\n",
      "Esophagus-Mucosa  variants ...\n",
      "(97676, 14)   to   (14344, 14)\n",
      "14344  total tests...  4390  pvalues were reduced to 1\n",
      "1471 \t gene level qval<=0.1\n",
      "3148 \t locus level qval<=0.1\n",
      "1327 \t qval<0.01\n",
      "\n",
      "Esophagus-Muscularis  variants ...\n",
      "(87608, 14)   to   (14727, 14)\n",
      "14727  total tests...  3013  pvalues were reduced to 1\n",
      "1374 \t gene level qval<=0.1\n",
      "3310 \t locus level qval<=0.1\n",
      "1370 \t qval<0.01\n",
      "\n",
      "Heart-LeftVentricle  variants ...\n",
      "(84416, 14)   to   (14248, 14)\n",
      "14248  total tests...  3118  pvalues were reduced to 1\n",
      "809 \t gene level qval<=0.1\n",
      "1840 \t locus level qval<=0.1\n",
      "713 \t qval<0.01\n",
      "\n",
      "Lung  variants ...\n",
      "(89448, 14)   to   (15023, 14)\n",
      "15023  total tests...  2917  pvalues were reduced to 1\n",
      "1412 \t gene level qval<=0.1\n",
      "3456 \t locus level qval<=0.1\n",
      "1474 \t qval<0.01\n",
      "\n",
      "Muscle-Skeletal  variants ...\n",
      "(84241, 14)   to   (14178, 14)\n",
      "14178  total tests...  2836  pvalues were reduced to 1\n",
      "1665 \t gene level qval<=0.1\n",
      "4156 \t locus level qval<=0.1\n",
      "1838 \t qval<0.01\n",
      "\n",
      "Nerve-Tibial  variants ...\n",
      "(88752, 14)   to   (14879, 14)\n",
      "14879  total tests...  2744  pvalues were reduced to 1\n",
      "2191 \t gene level qval<=0.1\n",
      "5583 \t locus level qval<=0.1\n",
      "2538 \t qval<0.01\n",
      "\n",
      "Skin-NotSunExposed  variants ...\n",
      "(89011, 14)   to   (14968, 14)\n",
      "14968  total tests...  3205  pvalues were reduced to 1\n",
      "1101 \t gene level qval<=0.1\n",
      "2609 \t locus level qval<=0.1\n",
      "1034 \t qval<0.01\n",
      "\n",
      "Skin-SunExposed  variants ...\n",
      "(89048, 14)   to   (14920, 14)\n",
      "14920  total tests...  2783  pvalues were reduced to 1\n",
      "1955 \t gene level qval<=0.1\n",
      "5042 \t locus level qval<=0.1\n",
      "2204 \t qval<0.01\n",
      "\n",
      "Thyroid  variants ...\n",
      "(89139, 14)   to   (14935, 14)\n",
      "14935  total tests...  2812  pvalues were reduced to 1\n",
      "2033 \t gene level qval<=0.1\n",
      "5110 \t locus level qval<=0.1\n",
      "2400 \t qval<0.01\n",
      "\n",
      "WholeBlood  variants ...\n",
      "(81939, 14)   to   (13755, 14)\n",
      "13755  total tests...  2834  pvalues were reduced to 1\n",
      "1243 \t gene level qval<=0.1\n",
      "3034 \t locus level qval<=0.1\n",
      "1247 \t qval<0.01\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess\n",
    "import collections\n",
    "import os\n",
    "\n",
    "## FDR correction After gene level adjustment\n",
    "SHORTEN = {\n",
    "    \"Artery-Aorta\":\"Artery A.\"     ,\n",
    "    \"Artery-Tibial\": \"Artery T.\",\n",
    "    \"Adipose-Subcutaneous\": \"Adipose S.\",    \n",
    "    \"Adipose-Visceral\":\"Adipose V.\",\n",
    "    \"Brain-Caudate\":\"Caudate\"   , \n",
    "    \"Brain-Cerebellum\":\"Cerebellum\",\n",
    "    \"Cells-Transformedfibroblasts\": \"Fibroblast\",\n",
    "    \"Esophagus-Mucosa\": \"E. Mucosa\",\n",
    "    \"Esophagus-Muscularis\":\"E Muscularis\",\n",
    "    \"Heart-LeftVentricle\":\"Ventricule\",\n",
    "    \"Lung\": \"Lung\",\n",
    "    \"Muscle-Skeletal\": \"Muscle\",\n",
    "    \"Nerve-Tibial\":\"Nerve\",\n",
    "    \"Skin-NotSunExposed\": \"Skin Unexposed\",\n",
    "    \"Skin-SunExposed\":\"Skin Leg\",\n",
    "    \"Thyroid\":\"Thyroid\",\n",
    "    \"WholeBlood\": \"Blood\"\n",
    "}\n",
    "\n",
    "tissu = sorted(SHORTEN.keys())\n",
    "#path = \"/storage/szfeupe/Runs/GTEx_estr/Analysis_by_Tissue/\"\n",
    "path = \"/storage/szfeupe/Runs/650GTEx_estr/Analysis_by_Tissue/\"\n",
    "\n",
    "def fdrcorrection(tissue):\n",
    "    print(tissue, ' variants ...')\n",
    "    Tell = os.system('mv '+path+tissue+'/PQValues '+path+tissue+'/old_PQValues')\n",
    "#Get most signif. variant by gene from linear reg STRs\n",
    "    LR1=pd.read_csv(path+ tissue+\"/Lin_Reg_Out\" , '\\t')\n",
    "    \n",
    "#Locus level\n",
    "    LR1['p.wald'].to_csv('pvalues.txt', sep='\\n', index=False)\n",
    "    Tell = subprocess.call(\"/home/szfeupe/projects/GTEX_eSTRs/gtex-estrs/Scripts/PostRegressionFixes/fdr-correct.r\")\n",
    "    Qval=pd.read_csv('/home/szfeupe/projects/GTEX_eSTRs/gtex-estrs/Scripts/PostRegressionFixes/qvalues.txt', sep=' ')\n",
    "    \n",
    "    LR1['llqvalue']=list(Qval['qvalue'])\n",
    "    LR1['llsignif']=list(Qval['significant'])\n",
    "\n",
    "#Gene level\n",
    "    LR0 = LR1.sort_values(\"p.wald\").groupby(\"gene\", as_index=False).first()     \n",
    "    print(LR1.shape, '  to  ', LR0.shape)\n",
    "\n",
    "    #Add counts tests by gene\n",
    "    counts=pd.DataFrame({'cts' : LR1.groupby([\"gene\"]).size()})    ## This is the count by genes\n",
    "    genes = list(LR0['gene'])\n",
    "    LR0['NTEST']= list(counts.loc[genes]['cts'])\n",
    "    \n",
    "    #Gene level adjustment\n",
    "    #(1) min_pval* #test\n",
    "    LR0['AD.pval']=LR0['p.wald']*LR0['NTEST']\n",
    "    #(2) if AD_pval>1 => AD_pval=1\n",
    "    LR0['AD.pval'][LR0['AD.pval']>1] = 1\n",
    "    \n",
    "    #Save pval in file and FDR correct\n",
    "    LR0['AD.pval'].to_csv('pvalues.txt', sep='\\n', index=False)\n",
    "    Tell = subprocess.call(\"/home/szfeupe/projects/GTEX_eSTRs/gtex-estrs/Scripts/PostRegressionFixes/fdr-correct.r\")\n",
    "        \n",
    "    #FDR corrected... add to dataframe\n",
    "    Qval=pd.read_csv('/home/szfeupe/projects/GTEX_eSTRs/gtex-estrs/Scripts/PostRegressionFixes/qvalues.txt', sep=' ')\n",
    "    LR0['qvalue']=list(Qval['qvalue'])\n",
    "    LR0['significant']=list(Qval['significant'])\n",
    "\n",
    "#Merging\n",
    "    merging=['gene','chrom','str.id','str.start','beta','beta.se','p.wald','llqvalue','llsignif']\n",
    "    LRP = pd.merge(LR1,LR0, on=merging, how='left')\n",
    "    \n",
    "#Header arrangement\n",
    "    Head=['gene','chrom','str.id','str.start','p.wald','llqvalue','llsignif','NTEST','qvalue','significant','beta','beta.se']\n",
    "    Out=LRP[Head]\n",
    "    Out.to_csv(path+tissue+'/PQValues', sep='\\t', index=False)\n",
    "\n",
    "    S=LR0['AD.pval']\n",
    "    print(len(S),' total tests... ', len(S[S>=1]) , ' pvalues were reduced to 1')\n",
    "    print(len(LRP[LRP['qvalue'] <=0.1]),'\\t gene level qval<=0.1')\n",
    "    print(len(LRP[LRP['llqvalue'] <=0.1]),'\\t locus level qval<=0.1')\n",
    "    print(len(LRP[LRP['llqvalue'] <=0.01]),'\\t qval<0.01\\n')\n",
    "    return()\n",
    "#\n",
    "#\n",
    "for T in tissu:                     \n",
    "    fdrcorrection(T)\n",
    "    \n",
    "\n",
    "####   After this Script, plot the qqplot and then FDR correction below    \n",
    "####   Next: We run the code to prepare for heritability analysis (STR+SNP)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WholeBlood  variants ...\n",
      "(209131, 12)   to   (14845, 12)\n",
      "14845\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/lib/python3.4/site-packages/ipykernel/__main__.py:25: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14845  total tests...  5015  pvalues were reduced to 1\n",
      "\n",
      "630 \t qval<=0.1\n",
      "462 \t qval<=0.05\n",
      "261 \t qval<0.01\n",
      "\n",
      "Cells-Transformedfibroblasts  variants ...\n",
      "(216353, 12)   to   (15362, 12)\n",
      "15362\n",
      "15362  total tests...  4885  pvalues were reduced to 1\n",
      "\n",
      "955 \t qval<=0.1\n",
      "703 \t qval<=0.05\n",
      "395 \t qval<0.01\n",
      "\n",
      "Muscle-Skeletal  variants ...\n",
      "(215860, 12)   to   (15275, 12)\n",
      "15275\n",
      "15275  total tests...  5546  pvalues were reduced to 1\n",
      "\n",
      "304 \t qval<=0.1\n",
      "205 \t qval<=0.05\n",
      "0 \t qval<0.01\n",
      "\n",
      "Lung  variants ...\n",
      "(225784, 12)   to   (15711, 12)\n",
      "15711\n",
      "15711  total tests...  5673  pvalues were reduced to 1\n",
      "\n",
      "378 \t qval<=0.1\n",
      "264 \t qval<=0.05\n",
      "0 \t qval<0.01\n",
      "\n",
      "Adipose-Subcutaneous  variants ...\n",
      "(225770, 12)   to   (15918, 12)\n",
      "15918\n",
      "15918  total tests...  5751  pvalues were reduced to 1\n",
      "\n",
      "376 \t qval<=0.1\n",
      "287 \t qval<=0.05\n",
      "115 \t qval<0.01\n",
      "\n",
      "Artery-Tibial  variants ...\n",
      "(221263, 12)   to   (15584, 12)\n",
      "15584\n",
      "15584  total tests...  5360  pvalues were reduced to 1\n",
      "\n",
      "524 \t qval<=0.1\n",
      "383 \t qval<=0.05\n",
      "196 \t qval<0.01\n",
      "\n",
      "Esophagus-Mucosa  variants ...\n",
      "(227378, 12)   to   (16111, 12)\n",
      "16111\n",
      "16111  total tests...  5582  pvalues were reduced to 1\n",
      "\n",
      "475 \t qval<=0.1\n",
      "330 \t qval<=0.05\n",
      "128 \t qval<0.01\n",
      "\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
