{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Load expression\n",
      "Load annotation\n",
      "Load strs regression\n",
      "Load snps regression\n",
      "Load SNPs\n",
      "Load STRs\n",
      "Restrict to STRs samples\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1111, 10)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Restrict to eSTR genes only\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "72"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import math\n",
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import random\n",
    "import shutil\n",
    "import sys\n",
    "import gzip\n",
    "\n",
    "def PROGRESS(msg, printit=True):\n",
    "    if printit: # false for some messages when not in debug mode\n",
    "        sys.stderr.write(\"%s\\n\"%msg.strip())\n",
    "        \n",
    "def MakeZScoreTable(vals):\n",
    "    try:\n",
    "        Zscore=vals['beta']/vals['beta.se']\n",
    "    except:\n",
    "        Zscore=None\n",
    "    return Zscore\n",
    "\n",
    "def WriteCorrTable(indexed_genotypes):\n",
    "    \"\"\" generate correlation table using normalized genotype\n",
    "      _1_ ... _n_\n",
    "    1| 1  ... C1n\n",
    "    .|    ...\n",
    "    n|Cn1 ... Cnn=1\n",
    "    \"\"\"\n",
    "    G=indexed_genotypes.transpose()\n",
    "    variants = list(G.columns)\n",
    "    CMat=[]\n",
    "    print ('\\t\\t\\t**', len(variants))\n",
    "    for V1 in variants:\n",
    "        COV=[]\n",
    "        for V2 in variants:\n",
    "            X=G[V1].replace('None', np.nan).astype(float)\n",
    "            Y=G[V2].replace('None', np.nan).astype(float)\n",
    "            #COV.append(X.corr(Y))\n",
    "            if X.corr(Y) is np.nan:    #### \n",
    "                COV.append(0.0) #For missing LD we assume non linear corr (undetermined LD)\n",
    "            else:\n",
    "                COV.append(X.corr(Y))\n",
    "        CMat.append(COV)\n",
    "    return pd.DataFrame(CMat,columns=variants, index=variants) \n",
    "\n",
    "def lookfor (x,p):\n",
    "    for i in range(1,len(p.index)):\n",
    "        if x in p.values[i][0]:\n",
    "            top = p.values[i][0]\n",
    "            score = p.values[i][2]\n",
    "            return i,top, score\n",
    "\n",
    "PATH = '/storage/szfeupe/Runs/650GTEx_estr/Analysis_by_Tissue/Adipose-Subcutaneous/'\n",
    "EXPRFILE = PATH+'Corr_Expr.csv'\n",
    "EXPRANNOTFILE = '/storage/resources/dbase/human/hg19/gencode_gene_annotations_hg19.csv'\n",
    "CHROM = 2\n",
    "if \"chr\" not in str(CHROM): CHROM=\"chr%s\"%CHROM\n",
    "DISTFROMGENE = 100000\n",
    "STRGTFILE = '/storage/szfeupe/Runs/650GTEx_estr/Genotypes/GenotypesNormalized.table'\n",
    "SNPGTFILE = '/storage/szfeupe/Runs/650GTEx_estr/SNP_Analysis/chr2.tab'\n",
    "OUTFILE = '/storage/szfeupe/Runs/650GTEx_estr/Analysis_by_Tissue/Adipose-Subcutaneous/HH/caviar.out2'\n",
    "REG_STRs = PATH+'/Lin_Reg_Out'\n",
    "REG_SNPs = PATH+'SNP_Analysis/Lin_Reg_Out'\n",
    "TMPDIR = PATH+'HH/caviar_temps/'\n",
    "if not os.path.exists(TMPDIR):\n",
    "    os.mkdir(TMPDIR)\n",
    "DEBUG =True\n",
    "ESTRGENESFILE=None\n",
    "\n",
    "\n",
    "# Load expression\n",
    "DEBUG =True\n",
    "PROGRESS(\"\\nLoad expression\", printit=DEBUG)\n",
    "expr = pd.read_csv(EXPRFILE)\n",
    "samples_to_keep = list(expr.index)\n",
    "# Load annotation\n",
    "PROGRESS(\"Load annotation\", printit=DEBUG)\n",
    "expr_annot = pd.read_csv(EXPRANNOTFILE)\n",
    "expr_annot.index = expr_annot[\"probe.id\"].values\n",
    "expr_annot = expr_annot.reindex(list(expr.columns))\n",
    "expr_annot = expr_annot.dropna() \n",
    "expr_annot = expr_annot[expr_annot[\"gene.chr\"] == CHROM]\n",
    "# Load strs Regression\n",
    "PROGRESS(\"\\nLoad strs regression\", printit=DEBUG)\n",
    "strs = pd.read_csv(REG_STRs, sep=\"\\t\")\n",
    "strs = strs.loc[strs['chrom']==CHROM]\n",
    "# Load snps regression\n",
    "PROGRESS(\"\\nLoad snps regression\", printit=DEBUG)\n",
    "snps = pd.read_csv(REG_SNPs, sep=\"\\t\")\n",
    "snps = snps.loc[snps['chrom']==CHROM]\n",
    "del snps['Unnamed: 0']\n",
    "#Load SNP genotypes\n",
    "PROGRESS(\"Load SNPs\", printit=DEBUG)\n",
    "snpgt = pd.read_csv(SNPGTFILE, sep=\"\\t\",low_memory=False)\n",
    "snpgt = snpgt.loc[snpgt['chrom']==CHROM]\n",
    "# Load STR genotypes\n",
    "PROGRESS(\"Load STRs\", printit=DEBUG)\n",
    "strgt = pd.read_csv(STRGTFILE, sep=\"\\t\")\n",
    "strgt = strgt.loc[strgt['chrom']==CHROM]\n",
    "# Restrict to STR samples\n",
    "PROGRESS(\"Restrict to STRs samples\", printit=DEBUG)\n",
    "str_samples = samples_to_keep\n",
    "expr = expr.reindex(str_samples)\n",
    "snpgt = snpgt[[\"chrom\",\"start\"] + str_samples]\n",
    "snpgt.index = list(snpgt[\"start\"].apply(lambda x: \"SNP_%s\"%int(x)))\n",
    "strgt = strgt[[\"chrom\",\"start\"] + str_samples]\n",
    "strgt.index = list(strgt[\"start\"].apply(lambda x: \"STR_%s\"%int(x)))\n",
    "# Load eSTR results\n",
    "PROGRESS(\"Restrict to eSTR genes only\", printit=DEBUG)\n",
    "if ESTRGENESFILE is not None:\n",
    "    estr_genes = pd.read_csv(ESTRGENESFILE, sep=\"\\t\")\n",
    "    Genes = estr_genes.loc[estr_genes['qvalue']<=0.1]['gene']  # estrs at 10%FDR\n",
    "    expr_annot = expr_annot.loc[expr_annot['gene.id'].isin(list(Genes))]\n",
    "print (expr_annot.shape)\n",
    "#open output files\n",
    "Errorfile = open(TMPDIR+\"/Errorfile.out\", 'w')\n",
    "OUT = open(OUTFILE, \"w\")\n",
    "OUT.write(\"\\t\".join(['CHROM', 'gene', 'num.STRs.in.top5', 'top_snp', 'top_snp_score', 'top_str', 'top.str.score'])+'\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "    # For each gene, get all cis-variants and the best STR\n",
    "    for i in range(expr_annot.shape[0]):\n",
    "        gene=expr_annot.index.values[i]\n",
    "        ensgene = expr_annot[\"gene.id\"].values[i]  #'ENSG00000215912.7'\n",
    "        genedir=TMPDIR+\"/%s\"%gene\n",
    "        if not os.path.exists(genedir):\n",
    "            os.mkdir(genedir)\n",
    "        clean_cmd='rm '+\n",
    "        os.system(clean_cmd)\n",
    "        PROGRESS(\"Getting data for %s\"%gene, printit=DEBUG)\n",
    "        start = expr_annot[\"gene.start\"].values[i]\n",
    "        end = expr_annot[\"gene.stop\"].values[i]\n",
    "    # Pull out cis SNPs\n",
    "        PROGRESS(\"Getting cis SNPs for %s\"%gene)\n",
    "        cis_snps = snps[(snps[\"str.start\"] >= (start-DISTFROMGENE)) & (snps[\"str.start\"] <= (end+DISTFROMGENE))]\n",
    "        #print cis_snps.shape , '###'\n",
    "        cis_snps = cis_snps.loc[cis_snps[\"str.start\"].isin(list(snpgt[\"start\"]))]  ###\n",
    "        cis_variants = cis_snps.loc[cis_snps['gene']==ensgene]\n",
    "        cis_variants=cis_variants.sort_values(by=\"p.wald\").head(n=100)\n",
    "        cis_variants.index = cis_variants[\"str.start\"].apply(lambda x: \"SNP_%s\"%int(x))\n",
    "        L=list(cis_variants.index)\n",
    "    # Pull out cis STR\n",
    "        PROGRESS(\"Getting most significant cis STR for %s\"%gene)\n",
    "        cis_strs = strs[strs[\"gene\"]==ensgene].sort_values(\"p.wald\")\n",
    "        if cis_strs.shape[0]==0:\n",
    "            PROGRESS(\"There are no STRs found for %s... Gene not in LR table\"%gene)\n",
    "            continue\n",
    "        else: \n",
    "            cis_strs.index = cis_strs[\"str.start\"].apply(lambda x: \"STR_%s\"%int(x)) \n",
    "            L0 = list(cis_strs.index)\n",
    "        #\n",
    "        cis_variants = pd.concat([cis_snps, cis_strs])\n",
    "    # Make z file data\n",
    "        Ztable = MakeZScoreTable(cis_variants[['beta','beta.se']])\n",
    "        if Ztable is None:\n",
    "            Errorfile.write(gene+\": Z score could not be calculated; beta.se is probably 0 or null\\n\")\n",
    "            continue\n",
    "        else:\n",
    "            Ztable.to_csv(genedir+'/ZFILE', sep='\\t',header=None)\n",
    "    # Make LD file\n",
    "        genotypes = snpgt.loc[L]\n",
    "        genotypes = pd.concat([genotypes, strgt.loc[L0] ])\n",
    "        del genotypes['chrom']\n",
    "        del genotypes['start']\n",
    "        CorrMatrix = WriteCorrTable(genotypes)\n",
    "        CorrMatrix.to_csv(genedir+'/LDFILE', sep='\\t',header=None, index=None)\n",
    "        PROGRESS(\"Matrix of corr was sent to file for %s\"%gene)        \n",
    "    #Run caviar\n",
    "        caviar_cmd = \"CAVIAR -l %s -z %s -o %s/caviar -c 1 -f 1 > %s\"%(genedir+\"/LDFILE\", genedir+\"/ZFILE\", genedir, genedir+\"/log\")\n",
    "        os.system(caviar_cmd)\n",
    "    #Output results\n",
    "    #Output results\n",
    "        if not os.path.exists(genedir+'/caviar_post'):\n",
    "            Errorfile.write(gene+\": CAVIAR did not run.\\n\\tERROR: Segmentation fault (core dumped) in log file\\n\")\n",
    "            continue\n",
    "        else:\n",
    "            post = pd.read_csv(genedir+'/caviar_post', sep=\"\\t\", header=None)\n",
    "            p = post.head(5)\n",
    "            num_str = len([x for x in list(p[0].values) if \"STR_\" in x])\n",
    "            if 'STR_' in p[0][0]:\n",
    "                topstr = p[0][0]\n",
    "                topstrscore = p.values[0][2]\n",
    "                I,topsnp, topsnpscore =lookfor('SNP_', post)\n",
    "            else:\n",
    "                topsnp = p[0][0]\n",
    "                topsnpscore = p.values[0][2]\n",
    "                I, topstr , topstrscore =lookfor('STR_',post)\n",
    "            OUT.write(\"\\t\".join([CHROM, gene, str(num_str), topsnp, str(topsnpscore), str(topstr), str(topstrscore)])+'\\n')\n",
    "            print(\"**\\t\".join([CHROM, gene, str(num_str), topsnp, str(topsnpscore), str(topstr), str(topstrscore), str(I)])+'\\n')\n",
    "\n",
    "    OUT.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(263, 12) ##SNPs#\n",
      "100 1 (101, 12)\n",
      "\t\t\t** 101\n",
      "(5,)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Matrix of corr was sent to file for ENSG00000184731.5\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "'NoneType' object is not iterable",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-19-893b6484200d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     77\u001b[0m             \u001b[0mtopsnp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mp\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     78\u001b[0m             \u001b[0mtopsnpscore\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 79\u001b[0;31m             \u001b[0mI\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtopstr\u001b[0m \u001b[0;34m,\u001b[0m \u001b[0mtopstrscore\u001b[0m \u001b[0;34m=\u001b[0m\u001b[0mlookfor\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'STR_'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mpost\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     80\u001b[0m         \u001b[0mOUT\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"\\t\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mCHROM\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgene\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnum_str\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mtopsnp\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtopsnpscore\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtopstr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtopstrscore\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mI\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m'\\n'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     81\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"\\t\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mCHROM\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgene\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnum_str\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mtopsnp\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtopsnpscore\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtopstr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtopstrscore\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mI\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m'\\n'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: 'NoneType' object is not iterable"
     ]
    }
   ],
   "source": [
    "    OUT.close()\n",
    "    O=0\n",
    "    OUT = open(OUTFILE, \"w\")\n",
    "    OUT.write(\"\\t\".join(['CHROM', 'gene', 'num.STRs.in.top5', 'top_snp', 'top_snp_score', 'top_str', 'top.str.score'])+'\\n')\n",
    "    # For each gene, get all cis-variants and the best STR\n",
    "    for i in range(0, expr_annot.shape[0]):\n",
    "        gene=expr_annot.index.values[i]\n",
    "        ensgene = expr_annot[\"gene.id\"].values[i]  #'ENSG00000215912.7'\n",
    "        genedir=TMPDIR+\"/%s\"%gene\n",
    "        if not os.path.exists(genedir):\n",
    "            os.mkdir(genedir)\n",
    "        clear_cmd = \"rm \"+genedir+'/*'\n",
    "        os.system(clear_cmd)\n",
    "        #PROGRESS(\"Getting data for %s\"%gene, printit=DEBUG)\n",
    "        start = expr_annot[\"gene.start\"].values[i]\n",
    "        end = expr_annot[\"gene.stop\"].values[i]\n",
    "    # Pull out cis SNPs\n",
    "        #PROGRESS(\"Getting cis SNPs for %s\"%gene)\n",
    "        cis_snps = snps[(snps[\"str.start\"] >= (start-DISTFROMGENE)) & (snps[\"str.start\"] <= (end+DISTFROMGENE))]\n",
    "        cis_snps = cis_snps.loc[cis_snps['gene']==ensgene]\n",
    "        cis_snps.index = cis_snps[\"str.start\"].apply(lambda x: \"SNP_%s\"%int(x))\n",
    "        print (cis_snps.shape , '##SNPs#')\n",
    "        cis_variants = cis_snps.loc[cis_snps[\"str.start\"].isin(list(snpgt[\"start\"]))]  ###        \n",
    "        #cis_variants = cis_snps.loc[cis_snps['gene']==ensgene]\n",
    "        cis_snps=cis_variants.sort_values(by=\"p.wald\").head(n=100)\n",
    "        #cis_snps.index = cis_snps[\"str.start\"].apply(lambda x: \"SNP_%s\"%int(x))\n",
    "        L=list(cis_snps.index)\n",
    "    # Pull out cis STR\n",
    "\n",
    "    # Pull out cis STR\n",
    "        #PROGRESS(\"Getting most significant cis STR for %s\"%gene)\n",
    "        cis_strs = strs[strs[\"gene\"]==ensgene].sort_values(\"p.wald\")\n",
    "        if cis_strs.shape[0]==0:\n",
    "            PROGRESS(\"There are no STRs found for %s... Gene not in LR table\"%gene)\n",
    "            continue\n",
    "        else: \n",
    "            cis_strs.index = cis_strs[\"str.start\"].apply(lambda x: \"STR_%s\"%int(x))\n",
    "            best_str_start = int(cis_strs[\"str.start\"].values[0])\n",
    "            L0 = list(cis_strs.index)\n",
    "        #\n",
    "        cis_variants = pd.concat([cis_snps, cis_strs])\n",
    "        print(len(L), len(L0), cis_variants.shape)\n",
    "    # Make z file data\n",
    "        Ztable = MakeZScoreTable(cis_variants[['beta','beta.se']])\n",
    "        if Ztable is None:\n",
    "            Errorfile.write(gene+\": Z score could not be calculated; beta.se is probably 0 or null\\n\")\n",
    "            continue\n",
    "        else:\n",
    "            Ztable.to_csv(genedir+'/ZFILE', sep='\\t',header=None)\n",
    "    # Make LD file\n",
    "        genotypes = snpgt.loc[L]\n",
    "        genotypes = pd.concat([genotypes, strgt.loc[L0] ])\n",
    "        del genotypes['chrom']\n",
    "        del genotypes['start']\n",
    "        CorrMatrix = WriteCorrTable(genotypes)\n",
    "        CorrMatrix.to_csv(genedir+'/LDFILE', sep='\\t',header=None, index=None)\n",
    "        PROGRESS(\"Matrix of corr was sent to file for %s\"%gene)        \n",
    "    #Run caviar\n",
    "        caviar_cmd = \"CAVIAR -l %s -z %s -o %s/caviar -c 1 -f 1 > %s\"%(genedir+\"/LDFILE\", genedir+\"/ZFILE\", genedir, genedir+\"/log\")\n",
    "        os.system(caviar_cmd)\n",
    "    #Output results\n",
    "        if not os.path.exists(genedir+'/caviar_post'):\n",
    "            Errorfile.write(gene+\": CAVIAR did not run.\\n\\tERROR: Segmentation fault (core dumped) in log file\\n\")\n",
    "            continue\n",
    "        else:\n",
    "            post = pd.read_csv(genedir+'/caviar_post', sep=\"\\t\", header=None)\n",
    "            post = post.sort_values(post.columns[2], ascending=False)\n",
    "            p = post.head(5)\n",
    "            print (p[0].shape)\n",
    "            \n",
    "            num_str = len([x for x in list(p[0].values) if \"STR_\" in x])\n",
    "            if 'STR_' in p[0][0]:\n",
    "                topstr = p[0][0]\n",
    "                topstrscore = p.values[0][2]\n",
    "                I,topsnp, topsnpscore =lookfor('SNP_', post)\n",
    "            else:\n",
    "                topsnp = p[0][0]\n",
    "                topsnpscore = p.values[0][2]\n",
    "                I, topstr , topstrscore =lookfor('STR_',post)\n",
    "            OUT.write(\"\\t\".join([CHROM, gene, str(num_str),topsnp,str(topsnpscore),str(topstr), str(topstrscore),str(I+1)])+'\\n')\n",
    "            print(\"\\t\".join([CHROM, gene, str(num_str),topsnp,str(topsnpscore),str(topstr), str(topstrscore),str(I+1)])+'\\n')\n",
    "            strsscores = post.loc[post[0].isin(cis_strs.index)][[0,2]]\n",
    "            strsscores['chrom']=[CHROM]*strsscores.shape[0]\n",
    "            strsscores['gene']= [gene]*strsscores.shape[0]\n",
    "            strsscores.columns = ['str','score', 'chrom', 'gene']\n",
    "            with open('strs_core'+CHROM, 'a') as k:\n",
    "                (strsscores[['chrom', 'gene','str','score']]).to_csv(k, header=False, index=False,sep='\\t')  \n",
    "        O=0+1                  \n",
    "        if O==6:\n",
    "            break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "p\n",
    "if 'STR_' in p[0][0]:\n",
    "    topstr = p[0][0]\n",
    "    topstrscore = p.values[0][2]\n",
    "    print(lookfor('SNP_', post))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>STR_126162</td>\n",
       "      <td>0.030105</td>\n",
       "      <td>0.012175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SNP_82865</td>\n",
       "      <td>0.025945</td>\n",
       "      <td>0.010493</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SNP_126890</td>\n",
       "      <td>0.021431</td>\n",
       "      <td>0.008667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SNP_126888</td>\n",
       "      <td>0.019044</td>\n",
       "      <td>0.007702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SNP_143559</td>\n",
       "      <td>0.016886</td>\n",
       "      <td>0.006829</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              0         1         2\n",
       "100  STR_126162  0.030105  0.012175\n",
       "0     SNP_82865  0.025945  0.010493\n",
       "1    SNP_126890  0.021431  0.008667\n",
       "2    SNP_126888  0.019044  0.007702\n",
       "3    SNP_143559  0.016886  0.006829"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p.reindex()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "None\n"
     ]
    }
   ],
   "source": [
    "#I, topstr , topstrscore =lookfor('STR_',post)\n",
    "print(lookfor('STR_',post))\n",
    "#1 SNP_82865 0.0104928"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#OLD LOOP\n",
    "    for i in range(expr_annot.shape[0]):\n",
    "        gene=expr_annot.index.values[i]\n",
    "        ensgene = expr_annot[\"gene.id\"].values[i]  #'ENSG00000215912.7'\n",
    "        genedir=TMPDIR+\"/%s\"%gene\n",
    "        if not os.path.exists(genedir):\n",
    "            os.mkdir(genedir)        \n",
    "        clear_cmd = \"rm \"+genedir+'/*'\n",
    "        os.system(clear_cmd)\n",
    "        PROGRESS(\"Getting data for %s\"%gene, printit=DEBUG)\n",
    "        start = expr_annot[\"gene.start\"].values[i]\n",
    "        end = expr_annot[\"gene.stop\"].values[i]\n",
    "    # Pull out cis SNPs\n",
    "        PROGRESS(\"Getting cis SNPs for %s\"%gene)\n",
    "        cis_snps = snps[(snps[\"str.start\"] >= (start-DISTFROMGENE)) & (snps[\"str.start\"] <= (end+DISTFROMGENE))]\n",
    "        cis_snps = cis_snps.loc[cis_snps['gene']==ensgene]\n",
    "        cis_snps.index = cis_snps[\"str.start\"].apply(lambda x: \"SNP_%s\"%int(x))\n",
    "        print cis_snps.shape , '##SNPs#'\n",
    "        cis_variants = cis_snps.loc[cis_snps[\"str.start\"].isin(list(snpgt[\"start\"]))]  ###        \n",
    "        #cis_variants = cis_snps.loc[cis_snps['gene']==ensgene]\n",
    "        cis_snps=cis_variants.sort_values(by=\"p.wald\").head(n=100)\n",
    "        #cis_snps.index = cis_snps[\"str.start\"].apply(lambda x: \"SNP_%s\"%int(x))\n",
    "        L=list(cis_snps.index)\n",
    "    # Pull out cis STR\n",
    "        PROGRESS(\"Getting all cis STR for %s\"%gene)\n",
    "        cis_strs = strs[strs[\"gene\"]==ensgene].sort_values(\"p.wald\")\n",
    "        if cis_strs.shape[0]==0 :\n",
    "            PROGRESS(\"There are no STRs found for %s... Gene not in LR table\"%gene)\n",
    "            continue\n",
    "        elif cis_snps.shape[0]<=1:\n",
    "            PROGRESS(\"There are no or not enough SNPs found for %s... Gene not in LR table\"%gene)\n",
    "            continue\n",
    "        else: \n",
    "            cis_strs.index = cis_strs[\"str.start\"].apply(lambda x: \"STR_%s\"%int(x)) \n",
    "            L0 = list(cis_strs.index)\n",
    "        #\n",
    "        cis_variants = pd.concat([cis_snps, cis_strs])\n",
    "        print len(L), len(L0), cis_variants.shape\n",
    "    # Make z file data\n",
    "        Ztable = MakeZScoreTable(cis_variants[['beta','beta.se']])\n",
    "        #print cis_variants[['beta','beta.se']]\n",
    "        if Ztable is None:\n",
    "            Errorfile.write(gene+\": Z score could not be calculated; beta.se is probably 0 or null\\n\")\n",
    "            continue\n",
    "        else:\n",
    "            Ztable.to_csv(genedir+'/ZFILE', sep='\\t',header=None)\n",
    "    # Make LD file\n",
    "        \n",
    "        genotypes = snpgt.loc[L]\n",
    "        genotypes = pd.concat([genotypes, strgt.loc[L0] ])\n",
    "        del genotypes['chrom']\n",
    "        del genotypes['start']\n",
    "        CorrMatrix = WriteCorrTable(genotypes)\n",
    "        CorrMatrix.to_csv(genedir+'/LDFILE', sep='\\t',header=None, index=None)\n",
    "        PROGRESS(\"Matrix of corr for %s LDFILE %s ZFILE %s \"%(gene, str(CorrMatrix.shape), str(Ztable.shape)))        \n",
    "    #Run caviar\n",
    "        caviar_cmd = \"CAVIAR -l %s -z %s -o %s/caviar -c 1 -f 1 > %s\"%(genedir+\"/LDFILE\", genedir+\"/ZFILE\", genedir, genedir+\"/log\")\n",
    "        os.system(caviar_cmd)\n",
    "    #Output results\n",
    "        PROGRESS(genedir)\n",
    "        if not os.path.exists(genedir+'/caviar_post'):\n",
    "            Errorfile.write(gene+\": CAVIAR did not run.\\n\\tERROR: Segmentation fault (core dumped) in log file\\n\")\n",
    "            continue\n",
    "        else:\n",
    "            post = pd.read_csv(genedir+'/caviar_post', sep=\"\\t\", header=None)\n",
    "            post = post.sort_values(post.columns[2], ascending=False)\n",
    "            p = post.head(5)\n",
    "            print 'top 5 ranked variants', p.shape\n",
    "            num_str = len([x for x in list(p[0].values) if \"STR_\" in x])\n",
    "            if 'STR_' in p[0][0]:\n",
    "                topstr = p[0][0]\n",
    "                topstrscore = p.values[0][2]\n",
    "                I,topsnp, topsnpscore =lookfor('SNP_', post)\n",
    "            else:\n",
    "                topsnp = p[0][0]\n",
    "                topsnpscore = p.values[0][2]\n",
    "                I, topstr , topstrscore =lookfor('STR_',post)\n",
    "            OUT.write(\"\\t\".join([CHROM, gene, str(num_str),topsnp,str(topsnpscore),str(topstr), str(topstrscore),str(I+1)])+'\\n')\n",
    "            strsscores = post.loc[post[0].isin(cis_strs.index)][[0,2]]\n",
    "            strsscores['chrom']=[CHROM]*strsscores.shape[0]\n",
    "            strsscores['gene']= [gene]*strsscores.shape[0]\n",
    "            strsscores.columns = ['str','score', 'chrom', 'gene']\n",
    "            with open('strs_core'+CHROM, 'a') as k:\n",
    "                (strsscores[['chrom', 'gene','str','score']]).to_csv(k, header=False, index=False,sep='\\t')  \n",
    "            O=0+1\n",
    "        #break\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
